solutions/instances/main.tf

```diff
@@ -76,20 +76,25 @@ locals {
     tag   = var.cloud_log_metrics_bucket_access_tag
   } : null
 
+  bucket_retention_configs = merge(
+    local.at_bucket_config != null ? { (local.at_cos_target_bucket_name) = var.at_cos_bucket_retention_policy } : null,
+    local.cloud_log_data_bucket_config != null ? { (local.cloud_log_data_bucket) = var.cloud_log_data_bucket_retention_policy } : null
+  )
+
   buckets_config = concat(
     local.archive_bucket_config != null ? [local.archive_bucket_config] : [],
     local.at_bucket_config != null ? [local.at_bucket_config] : [],
     local.cloud_log_data_bucket_config != null ? [local.cloud_log_data_bucket_config] : [],
     local.cloud_log_metrics_bucket_config != null ? [local.cloud_log_metrics_bucket_config] : []
   )
 
-  archive_rule = var.existing_at_cos_target_bucket_name == null ? {
+  archive_rule = length(coalesce(local.buckets_config, [])) != 0 ? {
     enable = true
     days   = 90
     type   = "Glacier"
   } : null
 
-  expire_rule = var.existing_at_cos_target_bucket_name == null ? {
+  expire_rule = length(coalesce(local.buckets_config, [])) != 0 ? {
```

Comment:
coalesce appears redundant here. The local is concat which always outputs a list.